File: packages/arcasha-router/src/shadow/shadow.ts
/**
 * @arcasha/router — Shadow Evaluation (EXP-0002F / 0003C.3)
 *
 * 全エキスパートを毎ステップ評価 (オラクル = シャドウ) し、フル情報フィードバックを
 * Router に供給する。compute は呼び出し側 (WS/WebGPU/任意バックエンド) が注入する。
 */

import type { EvalResult, ExpertInfo, Task } from '../core/types.js';
import { evaluateTask } from '../core/observation.js';

export interface Injection {
  type: 'latency' | 'capability';
  node: string;
  factor: number;
}

/** 全エキスパートを評価 (compute は呼び出し側のバックエンド + キャッシュ) */
export async function evaluateAll(
  experts: ExpertInfo[],
  task: Task,
  compute: (node: ExpertInfo, task: Task) => Promise<EvalResult>,
  inject?: Injection | null,
): Promise<Record<string, EvalResult>> {
  const out: Record<string, EvalResult> = {};
  for (const n of experts) {
    const raw = await compute(n, task);
    let score = raw.score;
    let latency = raw.latencyMs;
    if (inject?.type === 'capability' && inject.node === n.nodeId) {
      score = Math.round(score * inject.factor * 1000) / 1000;
    }
    if (inject?.type === 'latency' && inject.node === n.nodeId) {
      latency = Math.round(latency * inject.factor);
    }
    out[n.nodeId] = { nodeId: n.nodeId, text: raw.text, score, latencyMs: latency };
  }
  return out;
}

/** ルールベース評価で EvalResult を作る (キャッシュ可能な生成) */
export function evaluateWith(node: ExpertInfo, task: Task, text: string, latencyMs: number): EvalResult {
  return { nodeId: node.nodeId, text, score: evaluateTask(task.capability, text), latencyMs };
}

File: packages/arcasha-router-py/arcasha_router/shadow.py
"""Shadow evaluation — full-information feedback (EXP-0002F / 0003C.3)."""

import asyncio

from .observation import evaluate_task


async def evaluate_all(experts, task, compute, inject=None):
    """Evaluate all experts; ``compute`` is injected by the caller (WS/WebGPU/any)."""
    out = {}
    for n in experts:
        raw = await compute(n, task)
        score, latency = raw["score"], raw["latencyMs"]
        if inject and inject.get("type") == "capability" and inject.get("node") == n["nodeId"]:
            score = round(score * inject["factor"], 3)
        if inject and inject.get("type") == "latency" and inject.get("node") == n["nodeId"]:
            latency = round(latency * inject["factor"])
        out[n["nodeId"]] = {"nodeId": n["nodeId"], "text": raw.get("text", ""), "score": score, "latencyMs": latency}
    return out


def evaluate_with(node, task, text, latency_ms):
    return {"nodeId": node["nodeId"], "text": text, "score": evaluate_task(task["capability"], text), "latencyMs": latency_ms}


def run_async(coro):
    """Small helper to run the async shadow loop from sync code."""
    return asyncio.run(coro)
